services/matching.js

```diff
@@ -0,0 +1,156 @@
+ * @return {MapiRequest}
+ */
+Matching.getMatching = function(config) {
+  v.assertShape({
+    matchPath: v.required(
+      v.arrayOf(
+        v.shape({
+          coordinates: v.required(v.coordinates),
+          approach: v.oneOf('unrestricted', 'curb'),
+          radius: v.range([0, 50]),
+          isWaypoint: v.boolean,
+          waypointName: v.string,
+          timestamp: v.date
+        })
+      )
+    ),
+    profile: v.oneOf('driving-traffic', 'driving', 'walking', 'cycling'),
+    annotations: v.arrayOf(v.oneOf('duration', 'distance', 'speed')),
+    geometries: v.oneOf('geojson', 'polyline', 'polyline6'),
+    language: v.string,
+    overview: v.oneOf('full', 'simplified', 'false'),
+    steps: v.boolean,
+    tidy: v.boolean
+  })(config);
+
+  config.profile = config.profile || 'driving';
+
+  var matchPath = {
+    coordinates: [],
+    approach: [],
+    radius: [],
+    isWaypoint: [],
+    waypointName: [],
+    timestamp: []
+  };
+
+  /**
+   * An ordered array of object with coordinates and related properties.
+   * This might differ from the HTTP API as we have combined
+   * all the properties that depend on the order of coordinates into
+   * one object for ease of use.
+   *
+   * @typedef {Object} MatchPath
+   * @property {Array<number>} coordinates - An array containing (longitude, latitude).
+   * @property {'unrestricted'|'curb'} [approach="unrestricted"] - Used to indicate how requested routes consider from which side of the road to approach a waypoint.
+   * @property {number} [radius=5] - A number in meters indicating the assumed precision of the used tracking device.
+   * @property {boolean} [isWaypoint=true] - Whether this coordinate is waypoint or not. Note! the first and last coordinates will always have to be true.
+   * @property {string} [waypointName] - Custom names for waypoint used for the arrival instruction in banners and voice instructions.
+   * @property {date} [timestamp] - Unix timestamp corresponding the coordinate.
+   */
+  config.matchPath.forEach(function(obj) {
+    matchPath.coordinates.push(obj.coordinates[0] + ',' + obj.coordinates[1]);
+
+    // isWaypoint
+    if (obj.hasOwnProperty('isWaypoint') && obj.isWaypoint != null) {
+      matchPath.isWaypoint.push(obj.isWaypoint);
+    } else {
+      matchPath.isWaypoint.push(true); // default value
+    }
+
+    ['approach', 'radius', 'waypointName', 'timestamp'].forEach(function(prop) {
+      if (obj.hasOwnProperty(prop) && obj[prop] != null) {
+        matchPath[prop].push(obj[prop]);
+      } else {
+        matchPath[prop].push('');
+      }
+    });
+  });
+
+  ['coordinates', 'approach', 'radius', 'waypointName', 'timestamp'].forEach(
+    function(prop) {
+      // avoid sending params which are all `;`
+      if (
+        matchPath[prop].every(function(value) {
+          return value == '';
```

Comment:
Can we stick with the more explicit `===` by default, except maybe for the case of `x == null` (null/undefined)? I find myself wondering whether the intention with the `==` is to perform coercion or not.